src/main/java/net/minecraftforge/server/command/CommandTrack.java

```diff
@@ -47,38 +68,305 @@ public int getRequiredPermissionLevel()
         {
-            return Collections.singletonList("te");
+            super(ForgeTimeTracker.ENTITY_UPDATE);
+        }
+
+        @Override
+        public String getName()
+        {
+            return "entity";
+        }
+
+        @Override
+        public String getUsage(ICommandSender sender)
+        {
+            return "commands.forge.tracking.entity.usage";
+        }
+
+        @Override
+        protected ITextComponent buildTrackString(ICommandSender sender, ForgeTimings<Entity> data)
+        {
+            Entity entity = data.getObject().get();
+            if (entity == null)
+                return TextComponentHelper.createComponentTranslation(sender, "commands.forge.tracking.invalid");
+
+            BlockPos currentPos = entity.getPosition();
+            String world = getWorldName(entity.world.provider.getDimension());
+            double averageTimings = data.getAverageTimings();
+            String tickTime = (averageTimings > 1000 ? TIME_FORMAT.format(averageTimings / 1000) : TIME_FORMAT.format(averageTimings)) + getTimeSuffix(
+                    averageTimings);
+
+            return TextComponentHelper.createComponentTranslation(sender, "commands.forge.tracking.timingEntry", entity.getName(),
+                    world, currentPos.getX(), currentPos.getY(), currentPos.getZ(), tickTime);
+        }
+    }
+
+    private static class TrackResultsTileEntity extends TrackResultsBaseCommand<TileEntity>
+    {
+
+        public TrackResultsTileEntity()
+        {
+            super(ForgeTimeTracker.TILE_ENTITY_UPDATE);
+        }
+
+        @Override
+        public String getName()
+        {
+            return "te";
+        }
+
+        @Override
+        public String getUsage(ICommandSender sender)
+        {
+            return "commands.forge.tracking.te.usage";
+        }
+
+        @Override
+        protected ITextComponent buildTrackString(ICommandSender sender, ForgeTimings<TileEntity> data)
+        {
+            TileEntity te = data.getObject().get();
+            if (te == null)
+                return TextComponentHelper.createComponentTranslation(sender, "commands.forge.tracking.invalid");
+
+            String name = getTileEntityName(te);
+            BlockPos pos = te.getPos();
+
+            double averageTimings = data.getAverageTimings();
+            String tickTime = (averageTimings > 1000 ? TIME_FORMAT.format(averageTimings / 1000) : TIME_FORMAT.format(averageTimings)) + getTimeSuffix(
+                    averageTimings);
+            return TextComponentHelper
+                    .createComponentTranslation(sender, "commands.forge.tracking.timingEntry", name,
+                            getWorldName(te.getWorld().provider.getDimension()),
+                            pos.getX(), pos.getY(), pos.getZ(), tickTime);
+        }
+
+        private String getTileEntityName(TileEntity tileEntity)
+        {
+            String className = tileEntity.getClass().getSimpleName();
```

Comment:
Calculate this only if the registry id is null, or it is a waste or time and memory